man/sd_journal_get_fd.xml

```diff
@@ -201,20 +201,16 @@ else {
 
     <para><function>sd_journal_process()</function> and
     <function>sd_journal_wait()</function> return one of
-    <constant>SD_JOURNAL_NOP</constant>,
-    <constant>SD_JOURNAL_APPEND</constant> or
-    <constant>SD_JOURNAL_INVALIDATE</constant> on success or a
-    negative errno-style error code. If
-    <constant>SD_JOURNAL_NOP</constant> is returned, the journal did
-    not change since the last invocation. If
-    <constant>SD_JOURNAL_APPEND</constant> is returned, new entries
-    have been appended to the end of the journal. If
-    <constant>SD_JOURNAL_INVALIDATE</constant>, journal files were
-    added or removed (possibly due to rotation). In the latter event,
-    live-view UIs should probably refresh their entire display, while
-    in the case of <constant>SD_JOURNAL_APPEND</constant>, it is
-    sufficient to simply continue reading at the previous end of the
-    journal.</para>
+    <constant>SD_JOURNAL_NOP</constant>, <constant>SD_JOURNAL_APPEND</constant>, and
+    <constant>SD_JOURNAL_INVALIDATE</constant> on success or a negative errno-style
+    error code. If <constant>SD_JOURNAL_NOP</constant> is returned, the journal did
+    not change since the last invocation. If <constant>SD_JOURNAL_APPEND</constant>
+    is returned, new entries have been appended to the end of the journal. If
+    <constant>SD_JOURNAL_INVALIDATE</constant>, journal files were added or removed
+    (possibly due to rotation). <constant>SD_JOURNAL_INVALIDATE</constant> can usually
+    be handled identically to <constant>SD_JOURNAL_APPEND</constant>, because the
+    journal reading code automatically handles the removal and addition of journal
+    files.</para>
```

Comment:
I don't think such an app that "shows the whole journal" is realistic. Our current journal access functions don't provide any information about the file from which messages originate, so it'd have to manage files on its own, possibly use our code to read each file individually, and reimplement interleaving on its own. I don't see why anybody would want to bother.

Also, journals typically run into gigabytes, with millions of lines. There is very little point in displaying this whole thing live.

All journal viewers, graphical or not, operate the same way: read a bunch of lines, output them to screen, wait for new lines, push them out, trimming oldest lines when there's too much to show. Doing any kind of invalidation when journal files are removed is not necessary (or easy).

Reply:
`gnome-logs` is a tool that kinda is a tool like that. Not sure if it handles the event properly, but it really should, and the event exists for tools like that.

I am not sure why an app would need to know about which file a message originates from. It's not supposed to care. The idea is that implements a list store backed by the journal that kinda behaves like an infinite array into both directions. And whenever it gets the INVALIDATE event it should refresh the view, but remember the cursor of the entry currently displayed at the top, and then start displaying from there again. If it does that it will correctly follow journal files being added and removed, while also being stable in its viewport, and without having to know anything about journal files.

And no, I am not saying it has to show the *whole* thing live. It should just show what is in its current viewport, but do that live, so that any changes to that viewport are instantly visible. i.e. consider you scroll back to `gnome-logs` to some old stuff, which gets rotated while you look at it. It should then disappear from view too, and the next closest log data should be shown.

Reply:
or whether graphical or not, the point i am making is that stuff that has a "changeable viewport" (i.e. a scrollbar) and can do "live" updates nicely then it would be wise to refresh the screen fully on the invalidate event, and use the journal cursor as position to restart displaying from, so that things remain stable.

Reply:
But what is the benefit of disappearing the entry you are looking at if the file is unlinked from the filesystem? There is not problem whatsover with continuing to show that entry, and other entries from the same journal file. Making the display jump just because the daemon unlinked a file would be bad UX.

Re-reading the whole state when a file goes away would be quite bad UX too. It's just too slow to be useful, particularly in a graphical app which is expected to be always responsive. IMHO the way that this should be implemented would be watching the file list and responding to inotify events like the sd-journal code does. But SD_JOURNAL_INVALIDATE is just not enough information to do this efficiently.

As predicted, gnome-logs handles APPEND and INVALIDATE exactly the same :) :
https://gitlab.gnome.org/GNOME/gnome-logs/blob/master/src/gl-journal.c#L311

Reply:
> But what is the benefit of disappearing the entry you are looking at if the file is unlinked from the filesystem? There is not problem whatsover with continuing to show that entry, and other entries from the same journal file. Making the display jump just because the daemon unlinked a file would be bad UX.

Well, but it goes both ways: it's not just about removing journal files, also about adding new ones in. i.e. consider a journal-remote type scenario were new journal files are added in that contain older log data. In that case you get SD_JOURNAL_INVALIDATE too, it might really be worth displaying what they contain instantly too.

Reply:
> Re-reading the whole state when a file goes away would be quite bad UX too. It's just too slow to be useful, particularly in a graphical app which is expected to be always responsive. IMHO the way that this should be implemented would be watching the file list and responding to inotify events like the sd-journal code does. But SD_JOURNAL_INVALIDATE is just not enough information to do this efficiently.

Well, you wouldn't reread the "the whole stat". Just the stuff that is currently on screen. I mean, sd-journal tells you the cursor for each log event anyway, hence just remember what's currently at the top of the screen, and load exactly as much as necessary to fill the screen, but not more.